contracts/utils/ExpiringMetaCallForwarder.sol

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+import "./SelfMulticall.sol";
+
+/// @notice Contract that forwards expiring meta-calls to ERC2771 contracts
+/// that trust it
+/// @dev The term "meta-call" is used as a subset of "meta-tx" where `value`
+/// is zero and `to` is a contract.
+/// Adapted from OpenZeppelin's MinimalForwarder.sol.
+contract ExpiringMetaCallForwarder is EIP712, SelfMulticall {
+    using ECDSA for bytes32;
+    using Address for address;
+
+    struct ExpiringMetaCall {
+        address from;
+        address to;
+        bytes data;
+        uint256 expirationTimestamp;
+    }
+
+    /// @notice If the meta-call with hash is already executed
+    mapping(bytes32 => bool) public metaCallWithHashIsExecuted;
```

Comment:
Why do we need to restrict a meta-call from being executed more than once? To prevent usage on different chains?

Reply:
Say a token contract is the call forwarding target and someone signed a meta-call to send 10 tokens to someone